navigation/libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/milestone/Milestone.java

```diff
@@ -0,0 +1,101 @@
+package com.mapbox.services.android.navigation.v5.milestone;
+
+import com.mapbox.services.android.navigation.v5.NavigationException;
+import com.mapbox.services.android.navigation.v5.RouteProgress;
+
+/**
+ * Base Milestone statement. Subclassed to provide concrete statements.
+ *
+ * @since 0.4.0
+ */
+@SuppressWarnings("WeakerAccess") // Public exposed for creation of milestone classes outside SDK
+public abstract class Milestone {
+
+  private Builder builder;
+
+  public Milestone(Builder builder) {
+    this.builder = builder;
+  }
+
+  /**
+   * Milestone specific identifier as an {@code int} value, useful for deciphering which milestone invoked
+   * {@link MilestoneEventListener#onMilestoneEvent(RouteProgress, String, int)}.
+   *
+   * @return {@code int} representing the identifier
+   * @since 0.4.0
+   */
+  public int getIdentifier() {
+    return builder.getIdentifier();
+  }
+
+  /**
+   * A milestone can either be passed in to the {@link com.mapbox.services.android.navigation.v5.MapboxNavigation}
+   * object (recommended) or validated directly inside your activity.
+   *
+   * @param previousRouteProgress last locations generated {@link RouteProgress} object used to determine certain
+   *                              {@link TriggerProperty}s
+   * @param routeProgress         used to determine certain {@link TriggerProperty}s
+   * @return true if the milestone trigger's valid, else false
+   * @since 0.4.0
+   */
+  public abstract boolean validate(RouteProgress previousRouteProgress, RouteProgress routeProgress);
```

Comment:
What does `valid` means in this context? That the milestone is properly constructed, or that the milestone is happening? If the latter, I think I'd change `validate` with something else. Considering it's a boolean, maybe something like `milestone.isOccurring`?